### ISSUE TITLE
Структура приложения Angular

### ISSUE THREAD

**Issue by IgorKulishov** [g1er/Andrew#30]
__Приложение Angular__
1) файлы Angular деляться условно разделить на два типа:
* __компоненты__ (components) - отвечают за отображение информации на странице и нажатия кнопок и ввод информации на странице. Каждая "компонента" содержит в себе html, css, typescript  составляющие.
* __сервисы__ (services) - принимают и передают информацию удаленным серверам посредством REST API http обращений.
 
2) Компоненты и сервисы написаны на typescript и представляют из себя классы

3) В typescript существуют специальные функции декораторы, которые ставятся перед классами и определяют класс как компоненту или сервис.
Декоратор это встроенная функция Angular, перед которой стоит знак `@`.
* Если класс определен как компонента, то перед классом ставится декоратор `@Component()`
* Соответственно перед сервисами ставится декоратор `@Injectable()`
4) Так же как правило названия файлов говорят о том содержит ли файл компоненту или сервис:
* Компонента имеет название типа home-page.component.ts
* Сервис имеет название типа login.service.ts

**Comment by IgorKulishov** [comment]
Хорошо что начал разбираться. 
1. Да, на самом деле не всегда cli добавляет новую компоненту в модуль (пр умолчанию главный модуль называется app.module.ts, но название может быть и другим).
А) Как ты уже понял тебе на самом деле надо вручную импортировать компоненту в файл модуля (app.module.js по умолчанию) прописывать в declarations внутри @NgModule({}) декоратора:
```
import { HereisMyComponent } from './path/to/my/component/here-is-my-component.compinent.ts';
@NgModule({
  declarations: [
    AppComponent,
    HereisMyComponent
  ],
  imports: [
    BrowserModule,
    FormsModule,
    HttpModule
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
```
Б. однако внутри свойства bootstrap находится только основной модуль, который и "обвязывается" (от слова bootstrap дословно шнурки "шнуровать" приложение).

2. приложение состоит из множества компонент.
* Компоненты обязательно должны иметь "view" т.е. имеют обязательно template/templateUrl где определяется html часть которая выводиться на экран
* Основная компонента как правило имеет название AppComponent, которая находится в корне проекта и как правило находится внутри файла app.component.ts
* основная компонента включает в себя все остальные компоненты и обязательно указывается внутри в свойстве bootstrap, декоратора NgModule({}) модуля приложения.
* компоненты как правило имеет специальные функциональные принадлежности, могут входить друг в друга и взаимодействовать между собой и в целом напоминают кубики или конструктор "лего"
* компоненты могут быть умными (smart components - которые взаимодействуют с сервисами и как правило могут включать в себя простые компоненты) и простыми (dummy components - которые как правило имеют ограниченый набор ф-ционала и как правило входят в состав smart components).

**Comment by g1er** [comment]
Но в итоге мне пришлось вписать в bootstrap новый компонент. Это так и должно быть или у меня какой-то глюк?

**Comment by g1er** [comment]
Кстати, в бонус-уроках к основному курсу по Ангуляру идет курс по Тайпскрипту. Знать бы раньше, можно было тебе не тратить кучу времени на объяснения и рассказы тех или иных тем, а просто смотреть уроки и обсуждать непонятные моменты =)
Там еще есть много интересного. Например, уроки по ES6, NodeJs, RxJS

**Comment by IgorKulishov** [comment]
Молодец Бро, замечательный вопрос - я никогда так не пробовал 👍 
Дело в том что ты замечательным образом "нарушил" общепринятые правила Angular но нашел выход который так же работает - супер.
Итак по порядку сначала объясню как "обычно" принято и потом почему работает в твоем случае и в чем разница:
1) я создал свой бренч и так же создал pull request / merge request из своего бренча в твой бренч где я так же написал подробно что я там сделал https://github.com/g1er/Andrew/pull/31
Там все просто, дело в том что обычно принято иметь один главный компонент приложения (верний в ерархии) и все остальные компоненты вкладывать внутри темплейта главной компоненты.
Т.о. если посмотришь на мой запрос ты увидешь что я перебросил 
A) <app-my></app-my>
B) <app-my2></app-my2>
из `index.html` в `app.component.html` и оставляю только одну главную компоненту для обвязки в верхнем уровне 
```
  bootstrap: [AppComponent]
```

2) теперь почему у тебя работает.
Дело в том что ты сделал три компоненты в верхнем уровне соответственно их все три и надо обвязывать и указывать соответственно в bootstrap:
```
bootstrap: [AppComponent, MyComponent, My2Component]
```

**Comment by g1er** [comment]
Да, я потом так и понял, насчет иерархии. В арр-компоненте прописываются все остальные.
Я сейчас иду дальше, прохожу 4-5 уроки 2го блока. Буду стараться использовать свободные вечера, чтобы форсировать учебу. Если будут непонятны моменты, буду спрашивать

**Comment by IgorKulishov** [comment]
Хорошо, молодец! Желаю успеха!
Задавай вопросы, будем создавать Pull requests и расписывать объяснения в уровках (в отдельных уроках).

**Comment by g1er** [comment]
Сегодня смотрел урок на тему двусторонней связи "компонент-шаблон", где для работы нужен был FormsModule. Однако в app.module.ts в импортах его не оказалось. Пришлось самому разобраться, найти эту библиотеку и вручную прописать импорт. Это скорее не вопрос, а просто заметка о том, что некоторые вещи в ангуляре не настроены автоматически (хотя библиотека работы с формами, мне кажется, должна подключаться на автомате при создании проекта). Видать действительно разница в версиях влияет

**Comment by IgorKulishov** [comment]
Посмотри https://github.com/g1er/Andrew/issues/32
Я не все успел написать, но пока вкратце.
Позже расспишу, если не понятно.
